### PR TITLE
first MVCC support

### PR DESCRIPTION
**Pull request by c4pt0r**
1. Add MVCC to original interfaces (alway handles newest version).
2. Add some tests.  

TODO (I'll put them in next commits, to avoid a big PR):
1. Get/Scan specific version.
2. Compact

the structure of MVCC kv:

Key1 => 3 (newest version)
Key1_3 => ... (value of this version)
Key1_2 => ... (value of this version)
Key2 => 2
Key2_2 =>  Tombstone (this key is deleted in the version)
Key2_1 => ...
Key3 => 4
Key3_4 => ...

### CODE REVIEW

**Comment by c4pt0r** [comment]
@ngaut 
@siddontang

File: kv/kv.go [hunk 1]
```diff
@@ -13,6 +13,47 @@
...
+// EncodedKey represents encoded key in low-level storage engine.
+type EncodedKey []byte
+
+// Key represents high-level Key type
+type Key []byte
+
+// Next returns the next key in byte-order
+func (k Key) Next() Key {
+	// add \x0 to the end of key
+	return append(append([]byte(nil), []byte(k)...), 0)
```
Comment (siddontang) [inline kv/kv.go]: This should allocate slice twice? 

First allocing a slice with same size with k, then appending 0 also re-allocs another slice with size len(k) + 1 ? 

```
buf := make([]byte, len(k) + 1)
copy(buf, k)
return buf
```

maybe more efficient, I don't know.

Comment (c4pt0r) [inline kv/kv.go]: That makes sense.

File: store/localstore/kv_test.go [hunk 2]
```diff
@@ -168,6 +169,12 @@ func (s *testKVSuite) TestSeek(c *C) {
 
 	insertData(c, txn)
 
+	ss, _ := s.s.(*dbStore).db.GetSnapshot()
+	i := ss.NewIterator(nil)
+	for i.Next() {
+		log.Warn("!!!!", i.Key(), i.Value())
```
Comment (shenli) [inline store/localstore/kv_test.go]: Remove this line?

File: store/localstore/kv.go [hunk 3]
```diff
@@ -152,12 +153,22 @@ func (s *dbStore) tryConditionLockKey(tID int64, key string, snapshotVal []byte)
 		return errors.Trace(kv.ErrLockConflict)
 	}
 
-	currValue, err := s.db.Get([]byte(key))
+	metaKey := codec.EncodeBytes(nil, []byte(key))
+	currValue, err := s.db.Get(metaKey)
+	if err == kv.ErrNotExist || currValue == nil {
```
Comment (siddontang) [inline store/localstore/kv.go]: error2.ErrorEqual

File: store/localstore/local_version_provider.go [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+	mu              sync.Mutex
+	lastTimeStampTs int64
+	n               int64
+}
+
+// GetCurrentVer implements the VersionProvider's GetCurrentVer interface.
+func (l *LocalVersioProvider) GetCurrentVer() (kv.Version, error) {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	ts := (time.Now().UnixNano() / int64(time.Millisecond)) << 18
```
Comment (siddontang) [inline store/localstore/local_version_provider.go]: why 18 here? 
using a const name may be better.

File: store/localstore/local_version_provider.go [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+	"sync"
+	"time"
+
+	"github.com/pingcap/tidb/kv"
+)
+
+// LocalVersioProvider uses local timestamp for version.
+type LocalVersioProvider struct {
+	mu              sync.Mutex
+	lastTimeStampTs int64
```
Comment (siddontang) [inline store/localstore/local_version_provider.go]: lastTimeStamp is enough for this field ?

File: store/localstore/txn.go [hunk 6]
```diff
@@ -190,11 +191,16 @@ func (txn *dbTxn) doCommit() error {
...
 	// Check dirty store
+	curVer, _ := globalVerProvider.GetCurrentVer()
 	err := txn.each(func(iter iterator.Iterator) error {
+		metaKey := codec.EncodeBytes(nil, iter.Key())
+		// put dummy meta key, write current version
+		b.Put(metaKey, codec.EncodeUint(nil, curVer.Ver))
+		mvccKey := MvccEncodeVersionKey(iter.Key(), curVer)
 		if len(iter.Value()) == 0 { // Deleted marker
-			b.Delete(iter.Key())
+			b.Put(mvccKey, tombstone)
```
Comment (siddontang) [inline store/localstore/txn.go]: how about set tombstone in above Set, is it allowed to put tombstone as the valid value in Set?

File: kv/kv.go [hunk 7]
```diff
@@ -13,6 +13,47 @@
 
 package kv
 
+import "bytes"
+
+// EncodedKey represents encoded key in low-level storage engine.
+type EncodedKey []byte
```
Comment (qiuyesuifeng) [inline kv/kv.go]: Here why not use Key type but define a EncodedKey?

File: kv/kv.go [hunk 8]
```diff
@@ -13,6 +13,47 @@
...
+
+// Key represents high-level Key type
```
Comment (qiuyesuifeng) [inline kv/kv.go]: It is better to use comment end of .

File: kv/kv.go [hunk 9]
```diff
@@ -13,6 +13,47 @@
...
+}
+
+// Cmp returns the comparison result of two key, if A > B returns 1, A = B
+// returns 0, if A < B returns -1.
+func (k Key) Cmp(another Key) int {
+	return bytes.Compare(k, another)
+}
+
+// Cmp returns the comparison result of two key, if A > B returns 1, A = B
+// returns 0, if A < B returns -1.
```
Comment (qiuyesuifeng) [inline kv/kv.go]: `if A > B returns 1, else if A = B returns 0, else returns -1.`  seems better.

File: kv/kv.go [hunk 10]
```diff
@@ -13,6 +13,47 @@
...
+}
+
+// Next returns the next key in byte-order
+func (k EncodedKey) Next() EncodedKey {
+	return EncodedKey(bytes.Join([][]byte{k, Key{0}}, nil))
+}
+
+// VersionProvider provides increasing IDs.
+type VersionProvider interface {
+	GetCurrentVer() (Version, error)
```
Comment (qiuyesuifeng) [inline kv/kv.go]: Perhaps GetCurVersion or GetCurrVersion is better.

File: store/localstore/snapshot.go [hunk 11]
```diff
@@ -61,20 +96,69 @@ func (s *dbSnapshot) Release() {
...
+		// Get real key from metaKey
+		key, _, err := MvccDecode(metaKey)
+		if err != nil {
+			// It's not a valid metaKey, maybe overflow (other data).
+			it.valid = false
+			break
+		}
+		// Get kv pair.
+		val, err := it.s.Get(key)
+		if err != nil && err != kv.ErrNotExist {
```
Comment (siddontang) [inline store/localstore/snapshot.go]: error2.ErrorEqual

File: kv/kv.go (hunk 10 above)
Comment (ngaut) [inline kv/kv.go]: I prefer CurrentVersion :)

File: store/localstore/kv_test.go [hunk 12]
```diff
@@ -210,8 +216,13 @@ func (s *testKVSuite) TestInc(c *C) {
 	c.Assert(err, IsNil)
 	c.Assert(n, Equals, int64(100))
 
+	log.Info(key)
```
Comment (qiuyesuifeng) [inline store/localstore/kv_test.go]: Remove this line.

File: store/localstore/mvcc_test.go [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+	// insert test data
+	txn, err := t.s.Begin()
+	c.Assert(err, IsNil)
+	for i := 0; i < 5; i++ {
+		val := encodeInt(i)
+		err := txn.Set(val, val)
+		c.Assert(err, IsNil)
+	}
+	txn.Commit()
+
```
Comment (qiuyesuifeng) [inline store/localstore/mvcc_test.go]: Remove empty line.

File: kv/kv.go (hunk 7 above)
Comment (c4pt0r) [inline kv/kv.go]: We need to distinguish these two types. Key cannot be use as EncodedKey, although they are both []byte.

File: store/localstore/txn.go (hunk 6 above)
Comment (c4pt0r) [inline store/localstore/txn.go]: I'll use nil as tombstone.

File: kv/kv.go [hunk 14]
```diff
@@ -13,25 +13,68 @@
...
+// Next returns the next key in byte-order.
+func (k Key) Next() Key {
+	// add \x0 to the end of key
+	buf := make([]byte, len([]byte(k))+1)
+	copy(buf, []byte(k))
+	return buf
+}
+
+// Cmp returns the comparison result of two key, if A > B returns 1, A = B
+// returns 0, if A < B returns -1.
```
Comment (qiuyesuifeng) [inline kv/kv.go]: Comment `if A > B returns 1, else if A = B returns 0, else returns -1.`

Comment (siddontang) [inline kv/kv.go]: The result will be 0 if a==b, -1 if a < b, and +1 if a > b. 
See http://golang.org/pkg/bytes/#Compare

File: store/localstore/mvcc_test.go [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+	c.Assert(cnt1, Greater, cnt)
+}
+
+func (t *testMvccSuite) TestMvccNext(c *C) {
+	txn, _ := t.s.Begin()
+	it, err := txn.Seek(encodeInt(2), nil)
+	c.Assert(err, IsNil)
+	c.Assert(it.Valid(), IsTrue)
+	for it.Valid() {
+		log.Error("seeking", []byte(it.Key()))
```
Comment (qiuyesuifeng) [inline store/localstore/mvcc_test.go]: Remove this line?

File: store/localstore/mvcc_test.go [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+	c.Assert(len(v), Greater, 0)
+	txn.Commit()
+
+	cnt := 0
+	t.scanRawEngine(c, func(k, v []byte) {
+		cnt++
+	})
+	txn, _ = t.s.Begin()
+	txn.Set(encodeInt(0), []byte("v"))
+	v, err = txn.Get(encodeInt(0))
```
Comment (qiuyesuifeng) [inline store/localstore/mvcc_test.go]: Need add some check?

Comment (c4pt0r) [inline store/localstore/mvcc_test.go]: Done

**Comment by c4pt0r** [comment]
@qiuyesuifeng 
@ngaut 
@siddontang 
PTAL

**Comment by siddontang** [comment]
LGTM

**Comment by qiuyesuifeng** [comment]
LGTM

File: store/localstore/kv.go [hunk 17]
```diff
@@ -46,12 +46,14 @@ type storeCache struct {
...
-	globalID int64
-	mc       storeCache
+	globalID          int64
+	globalVerProvider kv.VersionProvider
+	mc                storeCache
 )
 
 func init() {
 	mc.cache = make(map[string]*dbStore)
+	globalVerProvider = &LocalVersioProvider{}
```
Comment (ngaut) [inline store/localstore/kv.go]: LocalVersioProvider -> LocalVersionProvider

**Pull request merged**